var/spack/repos/builtin/packages/singularity/package.py

```diff
@@ -5,29 +5,124 @@
-
-    # When installing as root, the copy has to run before chmod runs
-    def install(self, spec, prefix):
-        make('install', parallel=False)
+    version('3.1.1', '158f58a79db5337e1d655ee0159b641e42ea7435')
+
+    depends_on('go')
+    depends_on('libuuid')
+    depends_on('libgpg-error')
+    depends_on('squashfs', type='run')
+    depends_on('git', when='@develop')  # mconfig uses it for version info
+
+    # Where go traditionally thinks the source should be.
+    # Used as working_dir in "edit" step.
+    @property
+    def sylabs_dir(self):
+        return join_path(self.stage.path, 'src/github.com/sylabs/')
+
+    # Unpack the tarball as usual, then make a link to the traditional
+    # location.
+    def do_stage(self, mirror_only=False):
+        super(Singularity, self).do_stage(mirror_only)
+        if not os.path.exists(self.sylabs_dir):
+            tty.debug("Making symbolic link for singularity in {0}".format(
+                self.sylabs_dir))
+            mkdirp(self.sylabs_dir)
+            os.symlink(self.stage.source_path,
+                       join_path(self.sylabs_dir, 'singularity'))
+
+    # Hijack the edit stage to run mconfig.
+    def edit(self, spec, prefix):
+        with working_dir(self.build_directory):
+            configure = Executable('./mconfig --prefix=%s' % prefix)
+            configure()
+
+    # Set these for use by MakefilePackage's default build/install methods.
+    build_targets = ['-C', 'builddir', 'parallel=False']
+    install_targets = ['install', '-C', 'builddir', 'parallel=False']
+
+    def setup_environment(self, spack_env, run_env):
+        # Point GOPATH at the top of the staging dir for the build
+        # step.
+        spack_env.prepend_path('GOPATH', self.stage.path)
+
+    # `singularity` has a fixed path where it will look for
+    # mksquashfs.  If it lives somewhere else you need to specify the
+    # full path in the config file.  Fix the config file after it's
+    # installed.
+    @run_after('install')
+    def fix_mksquashfs_path(self):
+        prefix = self.spec.prefix
+        squash_path = join_path(self.spec['squashfs'].prefix.bin, 'mksquashfs')
+        filter_file(r'^# mksquashfs path =',
+                    'mksquashfs path = {0}'.format(squash_path),
+                    join_path(prefix.etc, 'singularity', 'singularity.conf'))
+
+    #
+    # Assemble a script that fixes the ownership and permissions of several
+    # key files, install it, and tty.warn() the user.
+    # HEADSUP: https://github.com/spack/spack/pull/10412.
+    #
+    def perm_script(self):
+        return 'spack_perms_fix.sh'
+
+    def perm_script_tmpl(self):
+        return "{0}.j2".format(self.perm_script())
+
+    def perm_script_path(self):
+        return join_path(self.spec.prefix.bin, self.perm_script())
+
+    def _build_script(self, filename, variable_data):
+        with open(filename, 'w') as f:
+            env = spack.tengine.make_environment(dirs=self.package_dir)
```

Comment:
Any chance of not having writable here? Or a previous script existing that we are overwriting (and would that be an issue?)

Speaking of - where do we clean up the script?

Reply:
Ah nevermind I see - the user needs to run it! Okay I'm going to pull the updates and test. I'm glad you dug into this, it's a much better structure than I knew how to do :)

Reply:
I'm opening the script for writing, so that's the `w`.  Later, I chmod it `555`, aka `r-xr-xr-x`, not writable.

We don't clean it up, it's there for the user to use if/when needed.

Reply:
Yeah I saw that after, you can disregard all these comments, I hadn't looked down a few lines when I wrote them.